runtime/autoload/provider/node.vim

```diff
@@ -41,14 +41,20 @@ function! provider#node#Detect() abort
   if exists('g:node_host_prog')
     return g:node_host_prog
   endif
-  let global_modules = get(split(system('npm root -g'), "\n"), 0, '')
-  if v:shell_error || !isdirectory(global_modules)
+  let npm_global_modules = get(split(system('npm root -g'), "\n"), 0, '')
+  " `yarn global dir` returns the root directory of the global modules
+  let yarn_global_modules = get(split(system('yarn global dir'), "\n"), 0, '') . '/node_modules'
+  let is_node_modules_dir = isdirectory(npm_global_modules) || isdirectory(yarn_global_modules)
+  if v:shell_error || !is_node_modules_dir
     return ''
```

Comment:
> v:shell_error must also be checked immediately after the npm command.

Then the yarn logic needs to move inside the if statement so we don't return an empty string, right?

Reply:
sure

Reply:
is there not a more generic way in nodejs-land to check if a freakin' module is available? There are numerous package managers ...

Reply:
> is there not a more generic way in nodejs-land to check if a freakin' module is available? There are numerous package managers ...

I don't think so or I'm not aware of any universal solution for this. But also the main package managers (covers 99%) are `npm` & `yarn`

What if I do it like this instead? any drawbacks or issues?

```diff
diff --git a/runtime/autoload/provider/node.vim b/runtime/autoload/provider/node.vim
index 7d4d5698a..8b8b0f901 100644
--- a/runtime/autoload/provider/node.vim
+++ b/runtime/autoload/provider/node.vim
@@ -22,6 +22,14 @@ function! s:is_minimum_version(version, min_major, min_minor) abort
     \         && str2nr(v_list[1]) >= str2nr(a:min_minor)))
 endfunction
 
+function! s:get_node_client_entry(global_modules_path, cli_bin) abort
+  let entry_path = glob(a:global_modules_path . a:cli_bin)
+  if v:shell_error || !isdirectory(a:global_modules_path) || empty(entry_path)
+    return ''
+  endif
+  return entry_path
+endfunction
+
 " Support for --inspect-brk requires node 6.12+ or 7.6+ or 8+
 " Return 1 if it is supported
 " Return 0 otherwise
@@ -41,20 +49,13 @@ function! provider#node#Detect() abort
   if exists('g:node_host_prog')
     return g:node_host_prog
   endif
-  let npm_global_modules = get(split(system('npm root -g'), "\n"), 0, '')
-  " `yarn global dir` returns the root directory of the global modules
-  let yarn_global_modules = get(split(system('yarn global dir'), "\n"), 0, '') . '/node_modules'
-  let is_node_modules_dir = isdirectory(npm_global_modules) || isdirectory(yarn_global_modules)
-  if v:shell_error || !is_node_modules_dir
-    return ''
-  endif
   if !s:is_minimum_version(v:null, 6, 0)
     return ''
   endif
   let cli_bin = '/neovim/bin/cli.js'
-  let entry_point = empty(glob(npm_global_modules . cli_bin))
-        \ ? glob(yarn_global_modules . cli_bin)
-        \ : glob(npm_global_modules . cli_bin)
+  let entry_point = s:get_node_client_entry(get(split(system('npm root -g'), "\n"), 0, ''), cli_bin)
+        " `yarn global dir` returns the root directory of the global modules
+        \ || s:get_node_client_entry(get(split(system('yarn global dir'), "\n"), 0, '') . '/node_modules', cli_bin)
   if !filereadable(entry_point)
     return ''
   endif
```

Reply:
Looks good, but pass the `'npm root -g'` command to `s:get_node_client_entry` then call system from there, so that the `v:shell_error` check is coupled with it more obviously.

Also I wouldn't depend on the falsey-ness of strings in VimL. Check empty() or compare `''` explicitly:

```vim
let entry_point = s:get_node_client_entry('npm root -g', cli_bin)
let entry_point = '' !=# entry_point ? entry_point : s:get_node_client_entry('yarn global dir', cli_bin)
```

Also the check for `empty(glob(npm_global_modules . cli_bin))` is now missing, which changes the logic. Was that intentional?

Reply:
> Also the check for `empty(glob(npm_global_modules . cli_bin))` is now missing, which changes the logic. Was that intentional?

That's already happening inside the `s:get_node_client_entry` function here

```vim
let entry_path = glob(a:global_modules_path . a:cli_bin)
if v:shell_error || !isdirectory(a:global_modules_path) || empty(entry_path)
  return ''
endif
return entry_path
```

Reply:
> Looks good, but pass the `'npm root -g'` command to `s:get_node_client_entry` then call system from there, so that the `v:shell_error` check is coupled with it more 

The only reason I didn't do this was that for `yarn` I need to concat with `/node_modules` 

```vim
s:get_node_client_entry(get(split(system('npm root -g'), "\n"), 0, ''), cli_bin)
" vs
s:get_node_client_entry(get(split(system('yarn global dir'), "\n"), 0, '') . '/node_modules', cli_bin)
```

Reply:
@justinmk what do you think about this?

```vim
function! s:get_node_client_entry(package_manager_cmd) abort
  let cli_bin = '/neovim/bin/cli.js'
  " `yarn global dir` returns the root directory of the global modules
  let yarn_extra_path = a:package_manager_cmd ==# 'yarn global dir' ? '/node_modules' : ''
  let entry_path = glob(get(split(system(a:package_manager_cmd), "\n"), 0, '') . yarn_extra_path . cli_bin)
  if v:shell_error || empty(entry_path) || !isdirectory(entry_path)
    return ''
  endif
  return entry_path
endfunction

function! provider#node#Detect() abort
  if exists('g:node_host_prog')
    return g:node_host_prog
  endif
  if !s:is_minimum_version(v:null, 6, 0)
    return ''
  endif
  let entry_point = s:get_node_client_entry('npm root -g')
  let entry_point = !empty(entry_point) ? entry_point : s:get_node_client_entry('yarn global dir')
  if !filereadable(entry_point)
    return ''
  endif
  return entry_point
endfunction
```

Not a fan of this 

```vim
let yarn_extra_path = a:package_manager_cmd ==# 'yarn global dir' ? '/node_modules' : ''
``` 

but not sure how to solve this.

Reply:
That's fine. 

Small nit: for such small scopes `package_manager_cmd` name is too verbose, `cmd` is enough.

Reply:
Will do the changes, rebase & push then 👍